src/d3a/models/market/market_structures.py

```diff
@@ -34,75 +35,97 @@ class Clearing:
 def my_converter(o):
     if isinstance(o, DateTime):
         return o.isoformat()
+    return None
 
 
-@dataclass
-class Offer:
-    id: str
-    time: datetime
-    price: float
-    energy: float
-    seller: str
-    original_offer_price: float = None
-    seller_origin: str = None
-    seller_origin_id: str = None
-    seller_id: str = None
-    energy_rate: float = field(init=False)
-    attributes: Dict = None
-    requirements: List[Dict] = None
-
-    def __post_init__(self):
-        self.id = str(self.id)
+class BaseBidOffer:
+    """Base class defining shared functionality of Bid and Offer market structures."""
+    def __init__(self, id: str, time: datetime, price: float, energy: float,
+                 attributes: Dict = None, requirements: List[Dict] = None):
+        self.id = str(id)
+        self.time = time
+        self.price = price
+        self.energy = energy
         self.energy_rate = self.price / self.energy
+        self.attributes = attributes
+        self.requirements = requirements
 
-    def update_price(self, price):
+    def update_price(self, price: float) -> None:
         self.price = price
         self.energy_rate = self.price / self.energy
 
-    def __repr__(self):
-        return ("<Offer('{s.id!s:.6s}', '{s.energy} kWh@{s.price}', '{s.seller} {rate}'>"
-                .format(s=self, rate=self.energy_rate))
-
-    def __str__(self):
-        return ("{{{s.id!s:.6s}}} [origin: {s.seller_origin}] "
-                "[{s.seller}]: {s.energy} kWh @ {s.price} @ {rate}"
-                .format(s=self, rate=self.energy_rate))
+    def update_energy(self, energy: float) -> None:
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
 
-    def to_json_string(self, **kwargs):
-        """Convert the Offer object into its JSON representation.
+    def to_json_string(self, **kwargs) -> str:
+        """Convert the Offer or Bid object into its JSON representation.
 
         Args:
             **kwargs: additional key-value pairs to be added to the JSON representation.
         """
-        offer_dict = deepcopy(asdict(self))
+        obj_dict = deepcopy(self.__dict__)
         if kwargs:
-            offer_dict = {**offer_dict, **kwargs}
+            obj_dict = {**obj_dict, **kwargs}
+
+        obj_dict["type"] = self.__class__.__name__
 
-        offer_dict["type"] = "Offer"
-        offer_dict.pop('energy_rate', None)
+        return json.dumps(obj_dict, default=my_converter)
 
-        return json.dumps(offer_dict, default=my_converter)
+    def _to_csv(self) -> Tuple:
+        rate = round(self.energy_rate, 4)
+        return (rate, self.energy, self.price,
+                self.seller if isinstance(self, Offer) else self.buyer)
```

Comment:
I would personally duplicate this code and put it both in Offer and Bid. I think it would be a bit more appropriate than having this `if` here.